File: Java/src/main/java/Solution801_900/Solution872.java
package Solution801_900;

import tool.TreeGenerate;
import tool.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author windows
 */
public class Solution872 {

    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        StringBuilder leafFromTree1 = getLeafFromTree(root1, new StringBuilder());
        StringBuilder leafFromTree2 = getLeafFromTree(root2, new StringBuilder());
        return leafFromTree1.toString().equals(leafFromTree2.toString());
    }

    private StringBuilder getLeafFromTree(TreeNode root, StringBuilder sb){
        if (root == null){
            return sb;
        }
        if (root.left == null && root.right == null) {
            sb.append(root.val + '#');
        }
        getLeafFromTree(root.left, sb);
        getLeafFromTree(root.right, sb);
        return sb;
    }




    public static void main(String[] args) {
        Solution872 solution872 = new Solution872();
        TreeNode tree1 = TreeGenerate.createTree(new Integer[]{3, 5, 1, 6, 2, 9, 8, null, null, 7, 14});
        TreeNode tree2 = TreeGenerate.createTree(new Integer[]{3, 5, 1, 6, 71, 4, 2, null, null, null, null, null, null, 9, 8});
        List<TreeNode> treeList1 = TreeGenerate.createTreeList(new Integer[]{3, 5, 1, 6, 2, 9, 8, null, null, 7, 14}, new ArrayList<>());
        List<TreeNode> treeList2 = TreeGenerate.createTreeList(new Integer[]{3, 5, 1, 6, 71, 4, 2, null, null, null, null, null, null, 9, 8}, new ArrayList<>());
        TreeGenerate.print(treeList1);
        TreeGenerate.print(treeList2);
        System.out.println(solution872.leafSimilar(tree1, tree2));
    }
}

File: C++/src/Solution801_900/Solution872.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution872.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

//Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

bool leafSimilar(TreeNode* root1, TreeNode* root2) {
  function<void(TreeNode*, vector<int>&)> dfs = [&](TreeNode* root, vector<int>& l){
    if (!root){
      return ;
    }
    dfs(root->left, l);
    dfs(root->right, l);
    if (!root->left && !root->right){
      l.push_back(root->val);
    }
  };
  vector<int> l1, l2;
  dfs(root1, l1);
  dfs(root2, l2);
  if (l1.size() !=l2.size()) return false;
  for (int i = 0; i < l1.size(); ++i) {
    if (l1[i] != l2[i]){
      return false;
    }
  }
  return true;
}